### PR TITLE
fix: datetime for defaultLocale set fa

### PR DESCRIPTION
**Pull request by datamweb**
When `toDateTimeString() ` is used. If `defaultLocale = 'fa'`.
The value of the datetime will be `UTF-8 string (33) "۲۰۲۲-۰۶-۱۲ ۲۲:۳۰:۳۵"`.
In this case, the database enters the information as `0000-00-00 00:0:00`.
Which causes a bug in the shield.
In this PR I tried to solve this problem.

note: This method can also solve the problem. `$time = new Time('now','','en-US') `

```
Data of Table "auth_groups_users":

+----+---------+-------+--------------------+
| id | user_id | group | created_at         |
+----+---------+-------+--------------------+
| 7  | 7       | user  | 0000-00-00 00:0... |
| 22 | 22      | user  | 0000-00-00 00:0... |
| 23 | 23      | user  | 2022-06-12 19:3... |
+----+---------+-------+--------------------+
```

### CODE REVIEW

**Comment by datamweb** [comment]
@kenjis I'm not sure this is the best solution, please review.
I will also send the **test** after confirmation.

**Comment by kenjis** [comment]
[toDateTimeString()](https://codeigniter4.github.io/CodeIgniter4/libraries/time.html#todatetimestring) returns localized datetime string. So it is not appropriate. 

Your solution is good.

**Comment by datamweb** [comment]
I tried to put the test in the appropriate file. Adding this test to other files(`AccessTokens.php` , `MagicLinkController.php` ...) seems to cause bloating.
However please let me know if you need to change.

File: tests/Authorization/AuthorizableTest.php [hunk 1]
```diff
@@ -295,4 +296,24 @@ public function testCanCascadesToGroupsWithWildcards(): void
 
         $this->assertTrue($this->user->can('admin.access'));
     }
+
+    // see https://github.com/codeigniter4/shield/pull/238
```
Comment (kenjis) [inline tests/Authorization/AuthorizableTest.php:300]:
```suggestion
    /**
     * @see https://github.com/codeigniter4/shield/pull/238
     */
```

Comment (datamweb) [reply tests/Authorization/AuthorizableTest.php]: Done.

File: tests/Authorization/AuthorizableTest.php [hunk 2]
```diff
@@ -295,4 +296,24 @@ public function testCanCascadesToGroupsWithWildcards(): void
...
+    public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
+    {
+        locale::setDefault('fa');
```
Comment (kenjis) [inline tests/Authorization/AuthorizableTest.php:305]:
```suggestion
        $currentLocale = Locale::getDefault();
        Locale::setDefault('fa');
```

File: tests/Authorization/AuthorizableTest.php [hunk 3]
```diff
@@ -295,4 +296,26 @@ public function testCanCascadesToGroupsWithWildcards(): void
...
+            'created_at' => Time::now()->format('Y-m-d H:i:s'),
+        ]);
+
+        $this->user->addGroup('admin');
+
+        $this->dontSeeInDatabase('auth_groups_users', [
+            'user_id'    => $this->user->id,
+            'group'      => 'admin',
+            'created_at' => '0000-00-00 00:00:00',
+        ]);
```
Comment (kenjis) [inline tests/Authorization/AuthorizableTest.php:319]: Add `Locale::setDefault($currentLocale);`.

Comment (datamweb) [reply tests/Authorization/AuthorizableTest.php:320]: Done.
I'm sorry to bother you so much for a small change.

File: tests/Authorization/AuthorizableTest.php [hunk 4]
```diff
@@ -295,4 +296,28 @@ public function testCanCascadesToGroupsWithWildcards(): void
...
+            'group'      => 'admin',
+            'created_at' => Time::now()->format('Y-m-d H:i:s'),
+        ]);
+
+        $this->user->addGroup('admin');
+
+        $this->dontSeeInDatabase('auth_groups_users', [
+            'user_id'    => $this->user->id,
+            'group'      => 'admin',
+            'created_at' => '0000-00-00 00:00:00',
```
Comment (kenjis) [inline tests/Authorization/AuthorizableTest.php:319]: Isn't `'0000-00-00 00:00:00'` MySQL specific?

Does this test fail before applying your fixes?

File: tests/Authorization/AuthorizableTest.php (hunk 3 above)
Comment (kenjis) [reply tests/Authorization/AuthorizableTest.php:320]: No problem. Thanks.

File: tests/Authorization/AuthorizableTest.php [hunk 5]
```diff
@@ -295,4 +296,28 @@ public function testCanCascadesToGroupsWithWildcards(): void
...
+    public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
+    {
+        $currentLocale = Locale::getDefault();
+        Locale::setDefault('fa');
+
+        $this->hasInDatabase('auth_groups_users', [
+            'user_id'    => $this->user->id,
+            'group'      => 'admin',
+            'created_at' => Time::now()->format('Y-m-d H:i:s'),
+        ]);
```
Comment (kenjis) [inline tests/Authorization/AuthorizableTest.php:312]: `hasInDatabase()` inserts a record:
https://codeigniter4.github.io/CodeIgniter4/testing/database.html?highlight=hasindatabase#hasindatabase-table-data
Why do you need to insert the data here?

**Comment by kenjis** [comment]
@datamweb The following test should work, but it does not work because of CI4 Time bug. https://github.com/codeigniter4/CodeIgniter4/pull/6116

```php
    public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
    {
        $currentLocale = Locale::getDefault();
        Locale::setDefault('fa');

        Time::setTestNow('March 10, 2017', 'America/Chicago');

        $this->user->addGroup('admin');

        $this->seeInDatabase('auth_groups_users', [
            'user_id'    => $this->user->id,
            'group'      => 'admin',
            'created_at' => '2017-03-10 00:00:00',
        ]);

        Locale::setDefault($currentLocale);
    }
```

File: tests/Authorization/AuthorizableTest.php (hunk 5 above)
Comment (datamweb) [reply tests/Authorization/AuthorizableTest.php:312]: In fact, I was just trying to check the correct record of the date in the database.
That's why I used it.

**Comment by datamweb** [comment]
> ```
>     public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
>     {
>         $currentLocale = Locale::getDefault();
>         Locale::setDefault('fa');
> 
>         Time::setTestNow('March 10, 2017', 'America/Chicago');
> 
>         $this->user->addGroup('admin');
> 
>         $this->seeInDatabase('auth_groups_users', [
>             'user_id'    => $this->user->id,
>             'group'      => 'admin',
>             'created_at' => '2017-03-10 00:00:00',
>         ]);
> 
>         Locale::setDefault($currentLocale);
>     }
> ```

@kenjis, So should I replace the test?

File: tests/Authorization/AuthorizableTest.php (hunk 5 above)
Comment (datamweb) [reply tests/Authorization/AuthorizableTest.php:312]: Oh my God.
This is a ridiculous mistake, **I'm sorry.**
**Thanks for the explanation.**

Comment (MGatner) [reply tests/Authorization/AuthorizableTest.php:312]: Not your fault @datamweb, I have always felt that `hasInDatabse()` was very poorly named.

**Comment by MGatner** [comment]
Let's hold off on this until we resolve the issue with the framework. The User Guide seems strongly in favor of `toDateTimeString()` for database inserts, so this might be the correct code and the bug is actually that the framework is localizing the result.

**Comment by datamweb** [comment]
> Let's hold off on this until we resolve the issue with the framework. The User Guide seems strongly in favor of `toDateTimeString()` for database inserts, so this might be the correct code and the bug is actually that the framework is localizing the result.

**Okya**. Please let me know if need to do something.

**Comment by kenjis** [comment]
@MGatner The bug in the framework was fixed in the `develop`, but we can't use it in this repository until it is released.
What do we do?

**Comment by MGatner** [comment]
This PR is still good, it is just the test case that needs the framework change, correct? Since the test case is covering an external bug I don't think it is necessary - we can proceed with the content change. If you want to leave the test in place with a conditional skip it could check the framework version on `CodeIgniter\CodeIgniter`.

**Comment by MGatner** [comment]
... Alternatively I can make the bugfix release of 4.2.1 today and we can use that.

**Comment by kenjis** [comment]
@datamweb Update the test, and let's move on.

```diff
--- a/tests/Authorization/AuthorizableTest.php
+++ b/tests/Authorization/AuthorizableTest.php
@@ -2,6 +2,7 @@
 
 namespace Tests\Authorization;
 
+use CodeIgniter\CodeIgniter;
 use CodeIgniter\I18n\Time;
 use CodeIgniter\Shield\Authorization\AuthorizationException;
 use CodeIgniter\Shield\Models\UserModel;
@@ -302,22 +303,23 @@ public function testCanCascadesToGroupsWithWildcards(): void
      */
     public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
     {
+        if (version_compare(CodeIgniter::CI_VERSION, '4.2.1', '<')) {
+            $this->markTestSkipped('This test does not work on CI v4.2.0 or before due to a bug.');
+        }
+
         $currentLocale = Locale::getDefault();
         Locale::setDefault('fa');
 
-        $this->hasInDatabase('auth_groups_users', [
-            'user_id'    => $this->user->id,
-            'group'      => 'admin',
-            'created_at' => Time::now()->format('Y-m-d H:i:s'),
-        ]);
+        Time::setTestNow('March 10, 2017', 'America/Chicago');
 
         $this->user->addGroup('admin');
 
-        $this->dontSeeInDatabase('auth_groups_users', [
+        $this->seeInDatabase('auth_groups_users', [
             'user_id'    => $this->user->id,
             'group'      => 'admin',
-            'created_at' => '0000-00-00 00:00:00',
+            'created_at' => '2017-03-10 00:00:00',
         ]);
+
         Locale::setDefault($currentLocale);
     }
 }
```

File: tests/Authorization/AuthorizableTest.php [hunk 6]
```diff
@@ -295,4 +297,29 @@ public function testCanCascadesToGroupsWithWildcards(): void
 
         $this->assertTrue($this->user->can('admin.access'));
     }
+
+    /**
+     * @see https://github.com/codeigniter4/shield/pull/238
+     */
+    public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
+    {
+        if (version_compare(CodeIgniter::CI_VERSION, '4.2.1', '<')) {
```
Comment (MGatner) [inline tests/Authorization/AuthorizableTest.php:306]:
```suggestion
        // @phpstan-ignore-next-line
        if (version_compare(CodeIgniter::CI_VERSION, '4.2.1', '<')) {
```

**Review by MGatner — CHANGES_REQUESTED** [review]
PHPStan doesn't like the constant being variable.

File: tests/Authorization/AuthorizableTest.php (hunk 6 above)
Comment (datamweb) [reply tests/Authorization/AuthorizableTest.php:307]: PHPStan doesn't like `// @phpstan-ignore-next-line` . also I doesn't like PHPStan ☺ .

**Comment by kenjis** [comment]
```diff
--- a/rector.php
+++ b/rector.php
@@ -20,6 +20,7 @@
 use Rector\Config\RectorConfig;
 use Rector\Core\ValueObject\PhpVersion;
 use Rector\DeadCode\Rector\ClassMethod\RemoveUnusedPromotedPropertyRector;
+use Rector\DeadCode\Rector\If_\UnwrapFutureCompatibleIfPhpVersionRector;
 use Rector\DeadCode\Rector\MethodCall\RemoveEmptyMethodCallRector;
 use Rector\DeadCode\Rector\StmtsAwareInterface\RemoveJustPropertyFetchForAssignRector;
 use Rector\EarlyReturn\Rector\Foreach_\ChangeNestedForeachIfsToEarlyContinueRector;
@@ -80,6 +81,11 @@
             __DIR__ . '/tests',
         ],
 
+        // check on constant compare
+        UnwrapFutureCompatibleIfPhpVersionRector::class => [
+            __DIR__ . '/tests/Authorization/AuthorizableTest.php',
+        ],
+
         // Ignore files that should not be namespaced to their folder
         NormalizeNamespaceByPSR4ComposerAutoloadRector::class => [
             __DIR__ . '/src/Helpers',
diff --git a/tests/Authorization/AuthorizableTest.php b/tests/Authorization/AuthorizableTest.php
index 3667f7c..0a8ea43 100644
--- a/tests/Authorization/AuthorizableTest.php
+++ b/tests/Authorization/AuthorizableTest.php
@@ -308,6 +308,7 @@ public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
             $this->markTestSkipped('This test does not work on CI v4.2.0 or before due to a bug.');
         }
 
+        // @phpstan-ignore-next-line
         $currentLocale = Locale::getDefault();
         Locale::setDefault('fa');
 
```

**Comment by MGatner** [comment]
4.2.1 is out, safe to roll back these workarounds

File: tests/Authorization/AuthorizableTest.php [hunk 7]
```diff
@@ -295,4 +297,30 @@ public function testCanCascadesToGroupsWithWildcards(): void
...
+    /**
+     * @see https://github.com/codeigniter4/shield/pull/238
+     */
+    public function testCreatedAtIfDefaultLocaleSetFaWithAddGroup(): void
+    {
+        // @phpstan-ignore-next-line
+        if (version_compare(CodeIgniter::CI_VERSION, '4.2.1', '<')) {
+            $this->markTestSkipped('This test does not work on CI v4.2.0 or before due to a bug.');
+        }
+
```
Comment (kenjis) [inline tests/Authorization/AuthorizableTest.php:310]:
```suggestion
```

**Comment by kenjis** [comment]
@datamweb Removing the `if` in the test case is one last job for this PR.
Can you?

**Comment by datamweb** [comment]
Hi @kenjis , I'm **sorry** for the delay. I was busy these few days.
**Thanks** for the review.

**Comment by kenjis** [comment]
Okay, all green!
Merged.

**Pull request merged**